numpy/core/src/common/cblasfuncs.c

```diff
@@ -188,6 +188,7 @@ _bad_strides(PyArrayObject *ap)
     int itemsize = PyArray_ITEMSIZE(ap);
     int i, N=PyArray_NDIM(ap);
     npy_intp *strides = PyArray_STRIDES(ap);
+    npy_intp *dims = PyArray_DIMS(ap);
 
     if (((npy_intp)(PyArray_DATA(ap)) % itemsize) != 0) {
```

Comment:
Do we need this condition for matmul too? Or does the ufunc machinery ensure this?

Reply:
I mean specifically the `((npy_intp)(PyArray_DATA(ap)) % itemsize) != 0` condition

Reply:
The ufunc flag NPY_ITER_ALIGNED handles this for us. I added a test and checked that 
- the unaligned ndarrays are copied in [`nditer_const.c:2914`](https://github.com/numpy/numpy/blob/v1.15.4/numpy/core/src/multiarray/nditer_constr.c#L2915) and
- when entering `@typ@_matmul` pointers are always aligned